lib/ui/assets.cc

```diff
@@ -0,0 +1,63 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/lib/ui/assets.h"
+
+#include "flutter/lib/ui/window/platform_configuration.h"
+#include "flutter/assets/asset_manager.h"
+#include "flutter/lib/ui/ui_dart_state.h"
+#include "third_party/tonic/dart_binding_macros.h"
+#include "third_party/tonic/dart_library_natives.h"
+#include "third_party/tonic/logging/dart_invoke.h"
+#include "third_party/tonic/typed_data/typed_list.h"
+#include "third_party/tonic/typed_data/dart_byte_data.h"
+#include "third_party/tonic/dart_library_natives.h"
+
+using tonic::DartInvoke;
+using tonic::DartPersistentValue;
+using tonic::ToDart;
+
+namespace flutter {
+
+void Assets::loadAssetBytes(Dart_NativeArguments args) {
+  UIDartState::ThrowIfUIOperationsProhibited();
+  Dart_Handle callback = Dart_GetNativeArgument(args, 1);
+  if (!Dart_IsClosure(callback)) {
+    Dart_SetReturnValue(args, tonic::ToDart("Callback must be a function"));
+    return;
+  }
+  Dart_Handle asset_name_handle = Dart_GetNativeArgument(args, 0);
+  uint8_t* chars = nullptr;
+  intptr_t asset_length = 0;
+  Dart_Handle result = Dart_StringToUTF8(asset_name_handle, &chars, &asset_length);
+  if (Dart_IsError(result)) {
+    Dart_PropagateError(result);
+    return;
+  }
+  std::string asset_name = std::string{reinterpret_cast<const char*>(chars),
+                                       static_cast<size_t>(asset_length)};
+
+  std::shared_ptr<AssetManager> asset_manager = UIDartState::Current()
+                                      ->platform_configuration()
+                                      ->client()
+                                      ->GetAssetManager();
+  std::unique_ptr<fml::Mapping> data =
+        asset_manager->GetAsMapping(asset_name);
+
+  if (data == nullptr) {
+    return;
+  }
+
+  Dart_Handle byte_buffer =
```

Comment:
DBC

Overall design probably needs review, but in the meantime consider creating external typed data here to avoid a copy if possible. See: https://github.com/flutter/engine/blob/master/lib/ui/painting/image_encoding.cc#L62

Reply:
This is something I looked at initially, but I wasn't sure if it was safe to using the data from a mapping object like that.

Reply:
Yeah, if the underlying memory region is mapped as read-only, then making this external typed data might have to wait on https://github.com/dart-lang/sdk/issues/42785 if returning `UnmodifiableUint8ListView` causes problems.

Reply:
Yeah, since I can only get a `const void *` to the underlying data, and that API accepts `void *`, per my basic understanding of C++ there is no real way to make that work - since writing to the typed data would presumably cause some sort of awful runtime exception.

Reply:
The const qualifier can be casted away. It is safe to wrap the read-only buffer in a Dart type that disallows writes. The issue is whether other bits of dart:ui and the framework can use that type.

Reply:
ahh TIL, maybe some sort of lint we have is preventing `const_cast`? At any rate, I guess that would have to wait for another day anyway

Reply:
So I needed to static_cast, then const_cast.